paasta_tools/paasta_maintenance.py

```diff
@@ -0,0 +1,367 @@
+            machine_id['hostname'] = hostname
+            machine_id['ip'] = gethostbyname(hostname)
+        machine_ids.append(machine_id)
+    return machine_ids
+
+
+def build_maintenance_schedule_payload(hostnames, start, duration, drain=True):
+    """Creates the JSON payload needed to (un)schedule maintenance on the specified hostnames.
+    :param hostnames: a list of hostnames
+    :param start: the time to start the maintenance, represented as number of nanoseconds since the epoch
+    :param duration: length of the maintenance window, represented as number of nanoseconds since the epoch
+    :param drain: boolean to note whether we are draining (True) the specified hosts or undraining (False) them
+    """
+    schedule = get_maintenance_schedule().json()
+    machine_ids = get_machine_ids(hostnames)
+
+    unavailability = dict()
+    unavailability['start'] = dict()
+    unavailability['start']['nanoseconds'] = int(start)
+    unavailability['duration'] = dict()
+    unavailability['duration']['nanoseconds'] = int(duration)
+
+    window = dict()
+    window['machine_ids'] = machine_ids
+    window['unavailability'] = unavailability
+
+    if schedule:
+        for existing_window in schedule['windows']:
+            for existing_machine_id in existing_window['machine_ids']:
+                if existing_machine_id in machine_ids:
+                    existing_window['machine_ids'].remove(existing_machine_id)
+                    if not existing_window['machine_ids']:
+                        schedule['windows'].remove(existing_window)
+        if drain:
+            windows = schedule['windows'] + [window]
+        else:
+            windows = schedule['windows']
+    elif drain:
+        windows = [window]
+    else:
+        windows = []
+
+    payload = dict()
+    payload['windows'] = windows
+
+    return payload
+
+
+def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+    """Loads the mesos-slave credentials from the specified file. These credentials will be used for all
+    maintenance API requests.
+    :param mesos_secrets: optional argument specifying the path to the file containing the mesos-slave credentials
+    """
+    with open(mesos_secrets) as data_file:
+        data = json.load(data_file)
+    username = data['principal']
+    password = data['secret']
+    return username, password
+
+
+def drain(hostnames, start, duration):
+    """Schedules a maintenance window for the specified hosts and marks them as draining.
+    :param hostnames: a list of hostnames
+    :param start: the time to start the maintenance, represented as number of nanoseconds since the epoch
+    :param duration: length of the maintenance window, represented as number of nanoseconds since the epoch
+    """
+    payload = build_maintenance_schedule_payload(hostnames, start, duration, drain=True)
+    client_fn = get_schedule_client()
+    print client_fn(method="POST", endpoint="", data=json.dumps(payload)).text
```

Comment:
given that you're not raising the exception in base_api (line 79), I think you might get caught out here if there is an exception in the request, the client_fn won't return anything.